### PR TITLE
Architecture/cleanup package 2

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Refactor Package Substates and various other refactors

## Details
* Remov SetPackageRoyaltyConfig instruction
* Separate Package blueprint royalty/access rules state into different partitions
* Separate Package Code and Schemas into different partitions
* Separate Package blueprint dependencies into different partitions
* Add BlueprintVersion

### For Internal Integrators
Package substates have been majorly overhauled.

### CODE REVIEW

**Review by iamyulong — COMMENTED** [review]
First batch of comments.

**Review by iamyulong — COMMENTED** [review]
Part 2

File: radix-engine/src/system/system.rs [hunk 1]
```diff
@@ -110,6 +120,93 @@ impl TryFrom<ObjectHandle> for ActorObjectType {
...
+    KeyValueStore {
+        collection_index: u8,
+        schema_ident: KeyValueStoreSchemaIdent,
+    },
+    Function {
+        ident: String,
+        schema_ident: FunctionSchemaIdent,
+    },
+    Event(String),
+}
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:159]: Alternatively, we can just pass `TypePointer` to `validate_payload_against_blueprint_schema`.

And type pointers can be obtained from method of `BlueprintInterface`, such as `BlueprintInterface::get_function_input_type(&self) -> TypePointer`.

File: radix-engine/src/system/system.rs [hunk 2]
```diff
@@ -139,28 +237,57 @@ where
...
-        &'s mut self,
-        payload: &Vec<u8>,
-        type_ref: &TypeRef,
-        blueprint_schema: &'s ScryptoSchema,
-        blueprint_id: BlueprintId,
-        instance_schema: &'s Option<InstanceSchema>,
-    ) -> Result<(), LocatedValidationError<ScryptoCustomExtension>> {
-        match type_ref {
-            TypeRef::Blueprint(index) => {
+    // TODO: Move into Package Package
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:240]: Can we remove the 2 TODOs above?

File: radix-engine/src/errors.rs [hunk 3]
```diff
@@ -271,13 +276,23 @@ pub enum CreateObjectError {
 
 #[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
 pub enum SystemModuleError {
+    BlueprintSchemaValidationError(BlueprintSchemaValidationError),
```
Comment (iamyulong) [inline radix-engine/src/errors.rs:279]: Can we move `BlueprintSchemaValidationError` to `SystemError` as there is no payload validation system module?

File: radix-engine/src/system/system.rs [hunk 4]
```diff
@@ -169,27 +296,279 @@ where
...
+                        ),
+                    )
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+
+    // TODO: Move into Package Package
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:315]: And this TODO as well.

File: radix-engine/src/system/system.rs [hunk 5]
```diff
@@ -139,28 +237,57 @@ where
...
-        payload: &Vec<u8>,
-        type_ref: &TypeRef,
-        blueprint_schema: &'s ScryptoSchema,
-        blueprint_id: BlueprintId,
-        instance_schema: &'s Option<InstanceSchema>,
-    ) -> Result<(), LocatedValidationError<ScryptoCustomExtension>> {
-        match type_ref {
-            TypeRef::Blueprint(index) => {
+    // TODO: Move into Package Package
+    pub fn validate_schema_pointer(
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:241]: Rename to `validate_payload`?

File: radix-engine/src/system/system.rs [hunk 6]
```diff
@@ -169,27 +296,279 @@ where
...
+        let blueprint_interface = PackageNativePackage::get_blueprint_default_interface(
+            blueprint_id.package_address.as_node_id(),
+            blueprint_id.blueprint_name.as_str(),
+            self.api,
+        )?;
+
+        let mut schema_pointers = Vec::new();
+
+        // TODO: Move this cache into actor
+        let mut schema_cache = SchemaCache::new();
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:331]: Can remove this if have a `get_schema(package_address, hash) -> Option<ScryptoSchema>` system API?

Comment (talekhinezh) [reply radix-engine/src/system/system.rs:331]: How?

File: radix-engine/src/system/system.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:159]: great idea!

File: radix-engine/src/system/system.rs (hunk 6 above)
Comment (iamyulong) [reply radix-engine/src/system/system.rs:247]: See https://github.com/radixdlt/radixdlt-scrypto/pull/1114.

By removing `SchemaCache`, we even improved the performance of Radiswap by 2.5%.

**Review by iamyulong — APPROVED** [review]
Looks really good to me! Thanks for working through the comments.

The only open question is if we'd like to move the `BlueprintDefinitionInit => BlueprintDefinition` computation off-chain.

**Pull request merged**